### PR TITLE
JAPI-00083: StateID from State only.

### PR DESCRIPTION
**Pull request by lrenn**
Perhaps not the best fix, but if for some reason the WS doesn’t return the state ID, but does return the state name, allow these methods to work and not throw an NPE.  This previously happened when checking an archived work unit.

I also considered setting the ID during the copy() method of WorkUnitInfo but then the state of the object wouldn't represent the true response.

## Type of change:
- [X] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).

## Checklist:
- [X] I have created a corresponding JIRA ticket for this submission
- [X] My code follows the code style of this project.
   - [ ] I have applied the Eclipse code-format template provided.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [X] I have read the HPCC Systems CONTRIBUTORS document (https://github.com/hpcc-systems/HPCC-Platform/wiki/Guide-for-contributors).
- [X] The change has been fully tested:
  - [X] I have performed unit tests to cover my changes.
  - [ ] I have performed system test and covered possible regressions and side effects.
  - [X] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [X] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.

## Testing:

Via the attached JUnit tests.

### CODE REVIEW

**Comment by rpastrana** [comment]
@lrenn can you open an issue against HPCC for this issue: "if for some reason the WS doesn’t return the state ID, but does return the state name..."

File: wsclient/src/main/java/org/hpccsystems/ws/client/platform/Workunit.java [hunk 1]
```diff
@@ -192,11 +192,8 @@ public static boolean isFailedState(String state) throws Exception
             WuStateNameMap.put("RUNNING", WUState.RUNNING);
     }
 
-    public static WUState translateWUState(String state) throws Exception
+    public static WUState translateWUState(String state)
     {
-        if (WuStateNameMap.size() <= 0)
```
Comment (rpastrana) [inline wsclient/src/main/java/org/hpccsystems/ws/client/platform/Workunit.java]: why was this block removed? If the static block above executes, we're never going to throw this exception. But if the code changes, this exception will let us quickly identify the issue.

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsWorkUnitsClient.java [hunk 2]
```diff
@@ -378,10 +379,15 @@ static public boolean isWorkunitComplete(WUState state)
...
     public static WUState getStateID(WorkunitInfo wu)
     {
-        if (wu != null)
-            return getStateID(wu.getStateID());
-        else
-            return WUState.UNKNOWN;
+        if (wu != null) {
+            if (wu.getStateID() != null) {
+                return getStateID(wu.getStateID());
+            } else if (wu.getState() != null && wu.getState().trim().length() != 0) {
```
Comment (rpastrana) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsWorkUnitsClient.java]: we should either trim in translateWUState() or make sure the argument in the translateWUstate call below is trimmed as well. As is, we might pass this condition (wu.getState().trim().length() != 0), but fail to translate due to leading or trailing spaces on an otherwise valid state string.

File: wsclient/src/test/java/org/hpccsystems/ws/client/HPCCWsWorkUnitsClientTest.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import org.hpccsystems.ws.client.platform.WUState;
+import org.hpccsystems.ws.client.platform.Workunit;
+import org.hpccsystems.ws.client.platform.WorkunitInfo;
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+
+
+public class HPCCWsWorkUnitsClientTest {
+    @Test
```
Comment (rpastrana) [inline wsclient/src/test/java/org/hpccsystems/ws/client/HPCCWsWorkUnitsClientTest.java]: these are good, can we add comments so we know what each test is doing?

File: wsclient/src/main/java/org/hpccsystems/ws/client/platform/Workunit.java (hunk 1 above)
Comment (lrenn) [inline wsclient/src/main/java/org/hpccsystems/ws/client/platform/Workunit.java]: I didn't think it was worth throwing an exception to catch a static block not executing, that's defined right above.  I would have then had to handle the exception by swallowing, returning UNKNOWN, or changing the sig to also throw.

Comment (lrenn) [inline wsclient/src/main/java/org/hpccsystems/ws/client/platform/Workunit.java]: I'll add to my test case to make sure the map isn't empty, but throwing doesn't make sense.

Comment (lrenn) [inline wsclient/src/main/java/org/hpccsystems/ws/client/platform/Workunit.java]: Actually, the tests I added already ensure the map isn't empty.

Comment (rpastrana) [inline wsclient/src/main/java/org/hpccsystems/ws/client/platform/Workunit.java]: ok, but the throws was removed, which could affect code that called this function, please make sure that callers of this function (and callers of the caller) are not throwing this exception anymore.

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsWorkUnitsClient.java [hunk 4]
```diff
@@ -382,7 +382,7 @@ public static WUState getStateID(WorkunitInfo wu)
         if (wu != null) {
             if (wu.getStateID() != null) {
                 return getStateID(wu.getStateID());
-            } else if (wu.getState() != null && wu.getState().trim().length() != 0) {
+            } else if (wu.getState() != null) {
```
Comment (rpastrana) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsWorkUnitsClient.java]: ok, so now it looks like you were only concerned about attempting to translate a string made up of white space, is that no longer a concern? Also, are valid states with leading, trailing white space not a concern? If they are, let's trim in the translate function.

Comment (lrenn) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsWorkUnitsClient.java]: I'm not concerned about a string made up of only whitespace.  It will return UNKNOWN just as if it was empty.  Adding the trim().length() check was my original mistake.

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsWorkUnitsClient.java [hunk 5]
```diff
@@ -378,10 +379,15 @@ static public boolean isWorkunitComplete(WUState state)
      */
     public static WUState getStateID(WorkunitInfo wu)
     {
-        if (wu != null)
-            return getStateID(wu.getStateID());
-        else
-            return WUState.UNKNOWN;
+        if (wu != null) {
+            if (wu.getStateID() != null) {
```
Comment (rpastrana) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsWorkUnitsClient.java]: @lrenn please either remove the curly brackets around the if and else if blocks, or move the open curly to a new line as well as the else if, and squash your commits into 1 so we can finalize this pull request. Thanks.

Comment (lrenn) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsWorkUnitsClient.java]: http://www.oracle.com/technetwork/java/javase/documentation/codeconventions-142311.html#449

https://google.github.io/styleguide/javaguide.html#s4.1-braces

I'll change the brace format, but most Java codebases use those standards.  As for squashing, that can be done from github when you merge, but I'll close this PR and submit a new one.

**Comment by lrenn** [comment]
Closing to squash the commits.

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsWorkUnitsClient.java (hunk 5 above)
Comment (rpastrana) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsWorkUnitsClient.java]: @lrenn as part of the Pull Request process you checked off the "My code follows the code style of this project. " but I noticed " I have applied the Eclipse code-format template provided." was not checked off. 
Please take a look at this link: https://github.com/hpcc-systems/HPCC-JAPIs#contributions
